File: ui/src/entities/photoshop.ts
import { PhotoshopListAPIResult } from "../api/photoshop/listPhotoshopFile";
import { appConfig } from "../config";

export type PhotoshopFileProps = {
  id: number;
  name: string;
  image_url: any;
  image_extension?: string;
  file_url: string;
  width: any;
  height: any;
  created_at: string;
  updated_at: any;
};

export class PhotoshopFile {
  props: PhotoshopFileProps;
  constructor(props: PhotoshopFileProps) {
    this.props = props;
  }

  get id() {
    return this.props.id;
  }

  get filename() {
    return this.props.name;
  }

  get imageUrl() {
    return `${appConfig.api.baeURL}/dist/${this.props.image_url}.png`;
  }

  get filepath() {
    return this.props.file_url;
  }

  get width() {
    return this.props.width;
  }

  toScale(
    maxWidth: number,
    maxHeight: number,
  ): { width: number; height: number } {
    if (this.width > this.height) {
      const scale = maxWidth / this.width;
      return { width: maxWidth, height: this.height * scale };
    } else {
      const scale = maxHeight / this.height;
      return { width: scale * this.width, height: maxHeight };
    }
  }

  get height() {
    return this.props.height;
  }

  static from_api_list(res: PhotoshopListAPIResult) {
    return res.data.map((data: any) => new PhotoshopFile(data));
  }
}

type PhotoshopElementProps = {
  id: number;
  photoshop_id: number;
  name: string;
  layer_id: string;
  text: string;
  xi: number;
  xii: number;
  yi: number;
  yii: number;
  width: number;
  height: number;
  is_group: boolean;
  group_id: number;
  level: number;
  kind: string;
  component_id: any;
  image_url: string;
  image_extension: string;
  created_at: string;
  updated_at: any;
};

export class PhotoshopElement {
  props: PhotoshopElementProps;

  constructor(props: any) {
    this.props = props;
  }

  get id() {
    return this.props.id;
  }

  get name() {
    return this.props.name;
  }

  get level() {
    return this.props.level;
  }

  get groupId() {
    return this.props.group_id || "0";
  }

  get isGroup() {
    return this.props.kind === "group";
  }

  get isBackground() {
    return false;
  }

  get layerId() {
    return this.props.layer_id;
  }

  get componentColor() {
    return `#${333}`;
  }

  get componentId() {
    return this.props.component_id;
  }

  get isComponent(): boolean {
    return Boolean(this.props.component_id);
  }

  static from_api_list(res: any) {
    return res.data.data.map((data: any) => new PhotoshopElement(data));
  }
}

export class ElementTree {
  children: ElementTree[];
  element?: PhotoshopElement;

  constructor(element?: PhotoshopElement, children = []) {
    this.children = children;
    this.element = element;
  }

  addChildren(e: ElementTree) {
    this.children.push(e);
  }

  get isComponent() {
    return this.element?.isComponent;
  }

  get isBackground() {
    return this.element?.isBackground;
  }

  get color() {
    return this.element?.componentColor;
  }

  get isDir() {
    return Boolean(this.element?.isGroup);
  }

  get name() {
    return this.element?.name;
  }

  get id() {
    return this.element?.id;
  }

  static layout(elements: PhotoshopElement[]): ElementTree {
    const root = new ElementTree();
    ElementTree._layout("0", root, elements);
    return root;
  }

  static _layout(
    groupId: string,
    tree: ElementTree,
    elements: PhotoshopElement[],
  ) {
    const inLevel = elements.filter((e) => e.groupId === groupId);
    if (inLevel.length === 0) return tree;
    const childrens = inLevel.map((e) => new ElementTree(e));
    childrens.forEach((e) => {
      tree.addChildren(e);
      ElementTree._layout(e.element.layerId, e, elements);
    });
  }
}

File: internal/entities/photoshop.go
package entities

import "time"

type Photoshop struct {
	ID             int32     `json:"id,omitempty"`
	Name           string    `json:"name,omitempty"`
	Filepath       string    `json:"filepath,omitempty"`
	FileExtension  string    `json:"file_extension,omitempty"`
	ImagePath      string    `json:"image_path,omitempty"`
	ImageExtension string    `json:"image_extension,omitempty"`
	Width          int32     `json:"width,omitempty"`
	Height         int32     `json:"height,omitempty"`
	CreatedAt      time.Time `json:"created_at,omitempty"`
}

type PhotoshopElement struct {
	ID             int32  `json:"id,omitempty"`
	Xi             int32  `json:"xi"`
	Xii            int32  `json:"xii"`
	Yi             int32  `json:"yi"`
	Yii            int32  `json:"yii"`
	LayerID        string `json:"layer_id,omitempty"`
	Width          int32  `json:"width"`
	Height         int32  `json:"height"`
	Kind           string `json:"kind,omitempty"`
	Name           string `json:"name,omitempty"`
	IsGroup        bool   `json:"is_group"`
	GroupId        int32  `json:"group_id"`
	Level          int32  `json:"level"`
	PhotoshopId    int32  `json:"photoshop_id,omitempty"`
	Image          string `json:"image,omitempty"`
	Text           string `json:"text,omitempty"`
	ImageExtension string `json:"image_extension,omitempty"`
	ComponentID    int32  `json:"component_id"`
}

type PhotoshopComponent struct {
	ID       int32              `json:"id,omitempty"`
	Elements []PhotoshopElement `json:"elements,omitempty"`
	Width    int32              `json:"width,omitempty"`
	Height   int32              `json:"height,omitempty"`
	Color    string             `json:"color,omitempty"`
	Type     string             `json:"type,omitempty"`
	Xi       int32              `json:"xi,omitempty"`
	Xii      int32              `json:"xii,omitempty"`
	Yi       int32              `json:"yi,omitempty"`
	Yii      int32              `json:"yii,omitempty"`
}
